### PR TITLE
[WIP][19.07] switch to Python3

### PR DESCRIPTION
**Pull request by champtar**
Hi all,

With fedora 31, `python` now points to `python3` and 19.07 doesn't build anymore.
Here some cherry-picks from master, not tested for now.
@nbd168 I'm cherry-picking wireless-regdb update, no idea if there is any impact
@yangbolu1991 do you know if https://github.com/openwrt/openwrt/commit/df0d555ea5a984ccfc2a493b63267a6e24d87c5b can be cherry-picked as-is
@ynezz everything else is yours
I haven't included scons migration & removal commits, if everything is now fixed in master maybe we can move scons in the packages feeds also for 19.07 ?

to be continued

### CODE REVIEW

**Comment by ynezz** [comment]
@yousong @neheb what do you think?

**Comment by neheb** [comment]
As far as I understand this, this is for the main repository where python/host is not available (that's in the packages feed).

I don't really like this. 19.07 has been branched for months with seemingly no RC in sight.

As far as Fedora 31 is concerned, python2 is still available with dnf. Probably makes more sense to fix these packages to use python2.

**Comment by champtar** [comment]
@neheb we should have an rc0 tomorrow morning (@jow- message on IRC)
I'm in favor of dropping python3, not delaying any rc or release
The only things to decide before release is if we move scons to the package feed, and if we switch it to python3 (can be done later)
```
19:02:43     champtar | Hi all, since Fedora 31, python now points to python3, openwrt master is fine but I cannot build 19.07, my questions are: 1) is it on someone's todo list ? 2) if not do you prefer a patch that just replace python with python2 or cherry-pick everything needed from master ?
19:03:41 KanjiMonster | champtar: probably cherry pick, if reasonably small
19:04:09        rr123 | install venv with python2 and build with venv activated?
19:05:54     champtar | For now i've patched all the python that I could find to python2, and in master it seems everything is migrated to python3
19:07:23 KanjiMonster | python2's EOL is planned in less than 2 months, so we should try to not require it in an upcoming release
19:07:29     champtar | So cherry-pick might not be that small, but I like the idea of removing python2 dependency for 19.07
19:08:21 KanjiMonster | so let me rephrase that, as long as it is not unreasonably huge, do the backport ;)
```

**Comment by yousong** [comment]
Moving 19.07 to python3 at this point is like opening a can of worm.  As far as dl_github_archive.py is concerned, we have at least 2 bugfix patches since the initial conversion.

The python2 dependency is only for building things, not running things at the target board.  I think it's not that bad we still use it in the build system.  It's like go1.4 is still needed to bootstraping newer versions of golang releases.

If the current prereq check does not ensure that python points to python2, let's add a bugfix commit to make that happen.  The change should be much smaller

**Comment by bobafetthotmail** [comment]
I think moving to Python3 is important for the future but it does not make sense for 19.07, it's very late already for its release and this is not an easy and painless change.

Just fix for pointing directly to python2, if the distro does not ship that anymore, then it's on the user to deal with it, it's not like older OpenWrt versions can be compiled without python2 either.

**Comment by champtar** [comment]
I agree requiring python2 for build only is not "that bad", but in 6 months ubuntu 20.04 and fedora 32 will not ship python2. There will be workaround, but for the main repo we are only talking about 5 packages or files, so I don't think the can of worm is that big for the main repo:
- package/firmware/layerscape/ls-rcw
- package/firmware/wireless-regdb
- scripts/dl_github_archive.py
- scripts/flashing/jungo-image.py
- tools/b43-tools/files/b43-fwsquash.py

For scons we can just move it to the packages feeds, still using python2, so people can slowly fix it there if wanted

cherry-picking now is also less maintenance in the near future

I'm running a full ath79 build with only python3 in the PATH and this branch, will tell you what fails

If I don't get any ACK until tonight (for now I see 3 NACK) I'll send a patch to explicitly use python2

**Comment by ynezz** [comment]
> for the main repo we are only talking about 5 packages or files, so I don't think the can of worm is that big for the main repo

Main repo is not an issue I would say, but the other feeds could be that worm of cans. As rc1 was already tagged in 88bff692a56f29c9af402a633e02eb537d983874 I would stay on conservative side and go with Python2 for this release. It would be really unfortunate to make this change during rc1->rc2 period, many downstream projects are already using 19.07 as a base for their development.

**Comment by champtar** [comment]
#2533
